Tools/peg_generator/pegen/c_generator.py

```diff
@@ -302,44 +287,62 @@ def visit_Opt(self, node: Opt) -> FunctionCall:
             comment=f"{node}",
         )
 
-    def visit_Repeat0(self, node: Repeat0) -> FunctionCall:
-        if node in self.cache:
-            return self.cache[node]
-        name = self.gen.artificial_rule_from_repeat(node.node, False)
-        self.cache[node] = FunctionCall(
+    def _generate_artificial_rule_call(
+        self,
+        node: Any,
+        prefix: str,
+        rule_generation_func: Callable[[], str],
+        return_type: Optional[str] = None,
+    ) -> FunctionCall:
+        node_str = f"{node}"
+        key = f"{prefix}_{node_str}"
+        if key in self.cache:
+            name = self.cache[key]
+        else:
+            name = rule_generation_func()
+            self.cache[key] = name
+
+        return FunctionCall(
             assigned_variable=f"{name}_var",
             function=f"{name}_rule",
             arguments=["p"],
-            return_type="asdl_seq *",
-            comment=f"{node}",
+            return_type=return_type,
+            comment=node_str,
+        )
+
+    def visit_Rhs(self, node: Rhs) -> FunctionCall:
+        if node.can_be_inlined:
+            return self.generate_call(node.alts[0].items[0])
+
+        return self._generate_artificial_rule_call(
+            node,
+            "rhs",
+            lambda: self.gen.artificial_rule_from_rhs(node),
+        )
+
+    def visit_Repeat0(self, node: Repeat0) -> FunctionCall:
+        return self._generate_artificial_rule_call(
+            node,
+            "repeat0",
+            lambda: self.gen.artificial_rule_from_repeat(node.node, False),
```

Comment:
Can we use the kwarg here so that it's clear what `True` or `False` is for? Same goes for the call below.

```suggestion
            lambda: self.gen.artificial_rule_from_repeat(node.node, is_repeat1=False),
```

Reply:
Totally agree with this change.

Maybe we should provide similar change in python_generator.py file?
Cache rework in general will be useful for python_generator too, I suppose.
But I don't know how to appropriate test such changes.

Reply:
Let's do that in a separate PR to keep the changes here minimum

Reply:
Of course, separate PR is a great choice